Close sidebar when resizing from non mobile to mobile

jorgefilipecosta opened the pull request:
## Description
This PR aims to close https://github.com/WordPress/gutenberg/issues/1535. It adds a new feature that automatically closes the sidebar when resizing from non-mobile to mobile sizes.

Thanks to a suggestion by @youknowriad this PR now adds redux-responsive. It simplified the implementation of the feature.

## How Has This Been Tested?
With sidebar open resize your window to a small size < 782. Verify that the sidebar automatically closes.
Open the sidebar see that you can open and close it in mobile sizes. Resize your window in small sizes with sidebar open and verify it does not close automatically. It only should close when going from big sizes > 782 to small sizes < 782.

## Notes
The layout component was refactored to be class. I was asking my self if this logic should be on layout component or sidebar component. I decided on layout component to be more general and later we may want to add more resize actions that do not affect just the sidebar.

youknowriad reviewed (commented):
In the past, I used this lib https://github.com/AlecAivazis/redux-responsive to track browser dimensions.

Do you think, it could simplify the logic and just update the `isEditorSidebarOpened` reducer to return the default value based on the browser size if the reducer value is "null"?

---
Alternatively, we could extract this logic to its own component or HoC.

```js
function Layout( { width } ) {
  console.log( width );
} 

export default withWindowSize( Layout )
```

jorgefilipecosta committed:
Added redux-resposive usage to implement automatic sidebar disabling when resizing to small sizes.

jorgefilipecosta commented:
Hi @youknowriad, I totally like the idea of using redux-responsive. It is one more dependency but in my option, it is worth it. And it will for sure be useful in other situations. I changed this PR to make usage of redux-responsive (and added it). Most of the code changes now are related with the refactor from function to class.

jasmussen commented:
Nice work. Works nicely for me. 

Can it also _restore_ or reopen the sidebar if you had it open on desktop before? It doesn't have to if it adds complexity, but if it's trivial might be a nice enhancement. That would make it behave like this:

- Start on desktop with sidebar open. Resize to small, sidebar closes, resize to large, sidebar opens again.
- Start on desktop with sidebar closed. Resize to small, sidebar stays closed, resize to large, sidebar stays closed.

youknowriad commented on editor/layout/index.js:
```diff
@@ -21,46 +22,55 @@ import VisualEditor from '../modes/visual-editor';
+class Layout extends Component {
+	componentWillReceiveProps( nextProps ) {
+		if ( nextProps.isSidebarOpened && nextProps.isMobile && ! this.props.isMobile ) {
+			this.props.toggleSidebar();
```
I wonder if we could achieve it with a Higher-order reducer instead.

```js
const myReducer = combineReducer( { ... } );

const myEnhancedReducer = ( state, action ) => {
  const newState = myReducer( state, action );
  
  if ( // check that isMobile and the previous state is not mobile using state and newSTate ) {  
     newState.preferences = {
        ...newState.preferences,
        isSidebarOpened: false, 
    };
 }

  return newState;
}
```

Do you think it's better this way? cc @aduth

jorgefilipecosta replied:
Hi @youknowriad, for now, a Higher-order reduce could be an option. But when implementing the more advanced logic proposed by @jasmussen, I'm not certain if a high order reducer makes things simpler. Because if we want to automatically open the sidebar again when increasing the size if it was opened before we decreased size we need some state information which I was thinking on adding to the state of layout component and just dispatch close /open sidebar when needed. With a hight order component, I think we need to add more information to the redux state which is not bad I think but I'm not certain if this information necessary for this logic should be there.

youknowriad replied:
I think if we want the proposed logic, we may need two reducer value. `isSidebarOpened` for desktop and `isSidebarOpenedOnMobile` which may simplify a lot of things :)

jorgefilipecosta replied:
Yes, that is a very nice way of handling the situation!

jorgefilipecosta replied:
Hi @youknowriad, to get a high order reducer here as it needs access to responsive and to preferences, it would need to be a top-level high order reducer, so it can access both responsive and preferences. It would be our first top-level high order reducer, is that right or I'm seeing something wrong?

youknowriad replied:
True, but the more I think about it, the more two reducer values seems good. which also solves the persistence issue, since we'd have to persist only one of these two preferences.

jorgefilipecosta replied:
Thank you for confirming my logic @youknowriad. Two reducer values sound like a good option for this case. But I think if we add an isSidebarOpenedMobile preference we still have the persistence issue. Because even the expected behavior of that ticket is for the sidebar to be closed on reload on mobile even if it was opened before. Our current store persists implementation stores all the preferences and we want it to be generic and not contain editor logic, so the best way to handle this problem I think is using the same logic we have now in PR https://github.com/WordPress/gutenberg/pull/3331 but switch to ignore isSidebarOpenedMobile instead of isSidebarOpened.

Also with this change, we would at least need to change the isSidebarOppened selector to check if we should use the normal sidebarOppened flag or the mobile one. In the toggling places, we would also need some logic to decide if we are toggling mobile sidebar or the normal one. I think this are the changes needed but I may be pursuing a more complex path.

jasmussen commented:
Just to clarify, you should only implement the logic of proposed if it doesn't add undue code complexity. I'm describing a rare edge case that's non destructive, so if the code is much cleaner and better without that logic I'd definitely defer to Riad.

jorgefilipecosta commented:
Hi @jasmussen, I understood, I just specifying a case where if we have the current implementation handling it may be simpler than in a hight order reducer. For us to decide on the best approach to use.

gziolo commented:
Let me propose an alternative solution. It probably have some drawbacks in terms of UI and code duplication, but it might solved all issues discussed so far. I was trying to find an implementation that would avoid using tweaking Redux stare for mobile. It turns out that we already use a patter where we display controls for the block in a differentplace for mobile devices and the rest. I thought why not take a similar approach with sidebar. If we would display the sidebar only for non-mobile users we could keep the current logic untouched. The thing is that we would have to provide a parallel UI for mobile. It seems like an overlay would fit the best to what we want to have here. Such approach will keep mobile completely independent from the toggle state of sidebar. I’m not sure if we would be happy to use overlay here, but honestly speaking what we have now looks this way anyway. The best part of having overlay is that you don’t need to persist it’s state as users expect to have it closed by default 😃

jorgefilipecosta commented:
Hi @gziolo, we would need some state information to know if we should show the sidebar overlay or not. This flag indicating if we show sidebar overlay is essentially a showSidebarMobile so I think your suggestion is similar to @youknowriad idea.
If this flag is saved in the redux state we need a store middleware to not persist it. If this flag is saved in the component state we don't need it, but it may be strange to save the normal sidebar flag in redux and to save the sidebar overlay on mobile.

gziolo commented on editor/reducer.js:
```diff
@@ -657,6 +658,13 @@ export function metaBoxes( state = defaultMetaBoxState, action ) {
 }
 
+// Create responsive reducer with lib default breakpoints excluding small where we are currently using 782.
+const responsive = createResponsiveStateReducer( {
```
Should we use variables imported from `.scss` file instead?

jorgefilipecosta replied:
Yes, that's the plan, but the PR that adds an scss importer https://github.com/WordPress/gutenberg/pull/3332 is not yet merged. It would be perfect if we can a new review on it and land it before this PR so we can make use of the importer :)

gziolo replied:
You referenced wrong PR number :)

jorgefilipecosta replied:
Yes, you are right, the correct PR number is https://github.com/WordPress/gutenberg/pull/3331. (comment was updated)

gziolo commented:
> Hi @gziolo, we would need some state information to know if we should show the sidebar overlay or not. This flag indicating if we show sidebar overlay is essentially a showSidebarMobile so I think your suggestion is similar to @youknowriad idea.
If this flag is saved in the redux state we need a store middleware to not persist it. If this flag is saved in the component state we don't need it, but it may be strange to save the normal sidebar flag in redux and not save the sidebar overlay flag.

Yes, it's a similar idea. We should have two different ways of controlling sidebar on mobile and desktop. Pick whatever you prefer in this case. I guess what @youknowriad proposed is easier to implement because it doesn't require changes in the UI thus moving additional complexity there. I think this is what @jasmussen wanted to avoid here. Saying that and reading:

>  so the best way to handle this problem I think is using the same logic we have now in PR #3331 but switch to ignore isSidebarOpenedMobile instead of isSidebarOpened.

I think that this proposal is a reasonable way of tackling this complex use case :)

jorgefilipecosta committed:
Updated store-defaults with the isSidebarOpenedMobile flag and removed old mobile logic; Updated disableIsSidebarOpenedOnMobile function part of mobile middleware to use the new isSidebarOpenedMobile flag;

jorgefilipecosta committed:
Implemented isMobile selector and added mobile logic to isEditorSidebarOpened selector.

jorgefilipecosta committed:
Updated toggleSidebar action and associated reducer with a new isMobile flag.

jorgefilipecosta commented:
Hi @gziolo, @youknowriad, I updated this PR to implement the two flags approach, creating the behavior described by @jasmussen. Feel free to have a new look.

gziolo commented on editor/components/block-settings-menu/block-inspector-button.js:
```diff
@@ -46,13 +46,21 @@ export function BlockInspectorButton( {
 export default connect(
 	( state ) => ( {
 		isSidebarOpened: isEditorSidebarOpened( state ),
+		isMobile: ! state.responsive.greaterThan.medium,
```
It would be nice to hide this logic behind `isMobile( state )` selector. I think this is a general pattern to always access `state` through selectors. @youknowriad can you confirm?

jorgefilipecosta replied:
I added an isMobile selector, it simplifies the logic and is more inline with the rest of our code base :)

gziolo commented on editor/reducer.js:
```diff
@@ -667,6 +679,15 @@ export function metaBoxes( state = defaultMetaBoxState, action ) {
 	}
 }
 
+// Create responsive reducer with lib default breakpoints excluding small where we are currently using 782.
```
I don't understand the part about 782, can you explain?

jorgefilipecosta replied:
I corrected the comment, the comment was valid before we had scss imports (added in parallel in other PR) and the values were hardcoded. But now it is not valid.

gziolo commented on editor/sidebar/header.js:
```diff
@@ -48,9 +49,16 @@ export default connect(
 	( state ) => ( {
 		panel: getActivePanel( state ),
 		count: getSelectedBlockCount( state ),
+		isMobile: ! state.responsive.greaterThan.medium,
```
We have 3 places where we need to pull `isMobile` value. Every time this value changes, React will detect that props have changed so it will try to re-render component even though `onToggleSidebar` method is not dynamically created. I'm wondering if this component will re-render when `isMobile` value changes with all those optimizations applied. It's a very interesting use case :)

gziolo commented on editor/store-defaults.js:
```diff
@@ -1,11 +1,7 @@
 export const PREFERENCES_DEFAULTS = {
 	mode: 'visual',
-	isSidebarOpened: ! viewPort.isExtraSmall(),
+	isSidebarOpened: true,
```
Nice :)

gziolo commented on editor/utils/mobile/index.js:
```diff
@@ -18,8 +6,8 @@ const isMobileChecker = () => 'object' === typeof window && window.innerWidth <
-export const disableIsSidebarOpenedOnMobile = ( payload, isMobile = isMobileChecker() ) => (
-	isMobile ? { ...payload, isSidebarOpened: false } : payload
+export const disableIsSidebarOpenedOnMobile = ( payload ) => (
+	payload.isSidebarOpenedMobile ? { ...payload, isSidebarOpenedMobile: false } : payload
```
👍

gziolo commented:
This looks very good, nice refactoring was done together with the behavior implemented 👍 

One more thing to consider. Would it make sense we move this whole `isMobile` detection to the middleware?
```js
export const mobileMiddleware = ( { getState } ) => next => action => {
	if ( action.type === 'REDUX_REHYDRATE' ) {
		// ...
	}
	if ( action.type === 'TOGGLE_SIDEBAR' ) {
		return next( {
			type: 'TOGGLE_SIDEBAR',
			isMobile: isMobile( getState() ),
		} );
	}
	return next( action );
};
```

Alternative solution with another action name:
```js
if ( action.type === 'TOGGLE_SIDEBAR' && isMobile( getState() ) ) {
	return next( {
		type: 'TOGGLE_SIDEBAR_MOBILE',
	} );
}
```

jorgefilipecosta committed:
Added functionality in mobile middleware to automatically add the is mobile flag in TOGGLE_SIDEBAR actions.

jorgefilipecosta commented:
Hi @gziolo, unfortunately, It did not occur to me using the mobile middleware to add the required flag but doing it simplifies a lot the logic. So I improved this PR with your suggestion. Thank you!

gziolo commented on editor/utils/mobile/index.js:
```diff
@@ -18,20 +7,26 @@ const isMobileChecker = () => 'object' === typeof window && window.innerWidth <
 		} );
 	}
+	if ( action.type === 'TOGGLE_SIDEBAR' ) {
+		return next( {
```
You can code it using `toggleSidebar( isMobile( getState() )` to take advantage of the existing `toggleSidebar` action creator.

gziolo commented on editor/utils/mobile/index.js:
```diff
@@ -1,15 +1,4 @@
- */
-const isMobileChecker = () => 'object' === typeof window && window.innerWidth < BREAK_MEDIUM;
-
+import { isMobile } from '../../selectors';
```
Missing docblock for `Internal dependencies`

gziolo commented on editor/reducer.js:
```diff
@@ -472,7 +481,10 @@ export function blockInsertionPoint( state = {}, action ) {
 	switch ( action.type ) {
 		case 'TOGGLE_SIDEBAR':
-			return {
+			return action.isMobile ? {
```
For my own curiosity, I tried a different syntax to avoid 2 objects :)
```js
const isSidebarOpenedKey = action.isMobile ? 'isSidebarOpenedMobile' : 'isSidebarOpened';
return {
    ...state,
    [ isSidebarOpenedKey ]: ! state[ isSidebarOpenedKey ],
};
```

It might be a bit harder to read, but sharing with you anyway :)

gziolo commented:
It looks great after all those iterations, thanks for working on it. I will test later today before I accept.

gziolo commented:
There is still one thing I'm not sure about. Steps to reproduce:
- open in mobile view
- open sidebar
- resize to desktop
- sidebar disappears
- resize to mobile again
- sidebar takes full screen

gziolo reviewed (approved):
Code looks good, there is one use case which is opened for discussion, but I wouldn't treat it as a blocker. Just something to keep in mind. Otherwise, it all works as expected!

jorgefilipecosta commented:
All the changes were applied @gziolo thank you for your efforts in reviewing.
I tried to replicate the behavior you described but I'm only able to replicate it if the sidebar is closed on the desktop.  So when if open sidebar on mobile I resize to the desktop the sidebar close than if I resize mobile it open again. I think this is the expected for now but in the future, we can, of course, try to improve some aspect :)
